connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/network/PreplayInternalResponse.kt

```diff
@@ -0,0 +1,28 @@
+package com.theoplayer.android.connector.uplynk.internal.network
+
+import com.theoplayer.android.connector.uplynk.network.PreplayResponse
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+
+private val json = Json{
+    ignoreUnknownKeys = true
+}
+
+@Serializable
+internal data class MinimalPreplayResponse(
+
+    /**
+     * The manifest's URL. (**NonNull**)
+     */
+    val playURL: String,
+
+    /**
+     * The identifier of the viewer's session. (**NonNull**)
+     */
+    val sid: String)
+
+
+internal class PreplayInternalResponse(val body: String) {
+    fun parseMinimalResponse(): MinimalPreplayResponse = json.decodeFromString(body)
+    fun parseExternalResponse(): PreplayResponse = json.decodeFromString(body)
```

Comment:
The MinimalPreplayResponse has only those fields that are needed to successfully start the playback. The PreplayResponse has all the fields that we have in documentation. 

My motivation was to not fail the playback in case we get some fields in PreplayResponse that we don't expect. In this case we send an ad error but not a fatal error.

This has advantages and disadvantages. Advantage is that our customers will be able to play video if we fail to parse full PreplayResponse. Disadvantage is that we hide the problem rather than fail hard.

What do you think, should we do like that or always parse full response?

Reply:
> The MinimalPreplayResponse has only those fields that are needed to successfully start the playback. The PreplayResponse has all the fields that we have in documentation.

The only difference between the two seems to be the addition of a `prefix` field. Which also seems pretty standard to me. I don't think it's worth splitting up.

> My motivation was to not fail the playback in case we get some fields in PreplayResponse that we don't expect.

Isn't that what `Json { ignoreUnknownKeys = true }` is for? ([docs](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-json/kotlinx.serialization.json/-json-configuration/ignore-unknown-keys.html))

Reply:
In the very near future it will be also `ads`, `drm` and `interstitialURL`. `ads` and `drm` are complex objects. Currently they are ignored because of `ignoreUnknownKeys = true` param, but if we get any parsing error inside those objects we will not play anything. If we ignore parsing errors in ads list we would still play video, but would have, for example, some ads not marked. And rather skip some unparsed Ad events.

Because the video is just a playback in my opinion it's too extreme to fail full playback when we just cannot parse one single ad.
 
I already see that their documentation is not complete and don't trust them 100%). Let me give and example:
Here is the citation from [these docs ](https://docs.edgecast.com/video/Content/Develop/Preplayv2.htm#AdObject)

> companions | Array | List of companion ads that go with the ad. **Companion ads are also ad objects.**

and they give an example:
```
"companions": [{
		    "adSlotID": "",
         	    "apiFramework": null,
                    ...
                }, {
                   "adSlotID": "",
                   "apiFramework": null,
               ...
               }
```
They didn't describe `adSlotID` in their docs for AdObject. It's not failing because of the `ignoreUnknownKeys = true` but I'm not sure what else did they miss in docs.

To me it looks better to have a bug from our client that some ad events are missing rather than some sources not able to be played. Taking into a count that Ads are embedded in the playback and their users will still see ads. 

The only drawback of this strategy is that we hide an issue if there is an issue rather than fail it hard. To mitigate this we could describe this in the documentation and advice them to throw an exception(or add analytics) in their error listener.

Reply:
I would still expect Uplynk to output conforming responses. 😛 But fair enough, it's probably better to try and play *something* than nothing. (Although I don't think we'll be able to do much if the `drm` field is invalid.)